runtime/domain.c

```diff
@@ -1736,9 +1737,12 @@ static void domain_terminate (void)
      sample at this point as the shared heap is gone. */
   caml_clear_gc_stats_sample(domain_state);
 
+  /* TODO: can this ever be NULL? can we remove this check? */
   if(domain_state->current_stack != NULL) {
     caml_free_stack(domain_state->current_stack);
   }
+  caml_free_backtrace_buffer(domain_state->backtrace_buffer);
+  caml_free_gc_regs_buckets(domain_state->gc_regs_buckets);
```

Comment:
This part is cleaning up some resource on the main domain thread and is not strictly speaking part of the systhread leak, isn't it?

Reply:
Yes indeed. (This is a separated fourth commit.) We thought it made more sense to employ the new resource-cleanup in both places where they initialized, for threads *and* for domains. (Otherwise the code ends up in an arguably inconsistent state.)

Reply:
+1 for cleaning up memory at domain exit too, not just thread exit.  The title of this PR and the Changes entry could reflect this fact.